exporter/exporterhelper/resource_to_label.go

```diff
@@ -0,0 +1,181 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package exporterhelper
+
+import (
+	"context"
+	"strconv"
+
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/consumer/pdata"
+)
+
+var (
+	resourceToLabels = &ResourceToLabels{}
+)
+
+// ResourceToLabels defines the consumer for converting resource attributes to labels
+type ResourceToLabels struct {
+	nextMetricsConsumer consumer.MetricsConsumer
+}
+
+// NewResourceToLabels creates a MetricsConsumer that converts the resource attributes to metric labels
+func NewResourceToLabels() consumer.MetricsConsumer {
+	return resourceToLabels
+}
+
+// ConsumeMetrics implements the consumer.ConsumeMetrics
+func (rtl *ResourceToLabels) ConsumeMetrics(ctx context.Context, md pdata.Metrics) error {
+	rms := md.ResourceMetrics()
+	for i := 0; i < rms.Len(); i++ {
+		resource := rms.At(i).Resource()
+		if resource.IsNil() {
+			continue
+		}
+		labelMap := extractLabelsFromResource(&resource)
+
+		ilms := rms.At(i).InstrumentationLibraryMetrics()
+		for j := 0; j < ilms.Len(); j++ {
+			ilm := ilms.At(j)
+			if ilm.IsNil() {
+				continue
+			}
+			metricSlice := ilm.Metrics()
+			for k := 0; k < metricSlice.Len(); k++ {
+				metric := metricSlice.At(k)
+				if metric.IsNil() {
+					continue
+				}
+				addLabelsToMetric(&metric, labelMap)
+			}
+		}
+	}
+	return rtl.nextMetricsConsumer.ConsumeMetrics(ctx, md)
+}
+
+// ConvertResourceToLabels converts all resource attributes to metric labels
+func ConvertResourceToLabels(md pdata.Metrics) {
+	rms := md.ResourceMetrics()
```

Comment:
Both these solutions are not correct right now, because one of the contract that the exporter needs to follow is to not change the incoming data (concurrency issues). So you have to create a copy of the data and then change the copy. An alternative is to have a forEach equivalent that accepts a Label set and resource, but that is also not efficient because you do conversion to string multiple times for resource attributes.

So right now the best I can think of is a helper method that from Resource returns a Label set and exporter have to manually call that and use both sets of labels when exporting. It is not ideal but best performance.

Reply:
Yea, make sense. I will modify the clone and update these options.

Reply:
Updated. Now it gets a clone of the metrics first and then modifies it.